### ISSUE TITLE
HTTP client uses SOCKS4 proxy hostname for SSL hostname verification

### ISSUE THREAD

**Issue by sgrimm** [eclipse/jetty.project#4366]
**Jetty version**
9.4.22.v20191022

**Java version**
11.0.5

**OS type/version**
OS X 10.15.1

**Description**
When the HTTP client is configured to use a SOCKS4 proxy, it passes the proxy hostname, not the destination hostname, to the SSL engine to do hostname verification. This didn't happen under 9.4.11.v20180605; I discovered the problem while upgrading from that version of the library.

Example code:

```java
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.Socks4Proxy;
import org.eclipse.jetty.client.api.ContentResponse;
import org.eclipse.jetty.util.ssl.SslContextFactory;

class Scratch {

  public static void main(String[] args) throws Exception {
    SslContextFactory sslContextFactory = new SslContextFactory.Client();
    // Or for 9.4.11
    // SslContextFactory sslContextFactory = new SslContextFactory();

    HttpClient client = new HttpClient(sslContextFactory);
    Socks4Proxy proxy = new Socks4Proxy("localhost", 12345);
    client.getProxyConfiguration().getProxies().add(proxy);

    client.start();

    ContentResponse response = client.GET("https://apple.com/");
    System.out.println(String.format("Got %d bytes", response.getContent().length));
  }
}
```

Start a SOCKS proxy on port 12345, e.g., by running `ssh -D 12345 localhost`, and run the code. It should print the number of bytes on the Apple home page (and does so under 9.4.11), but instead bombs out with an exception:

```
javax.net.ssl.SSLHandshakeException: No subject alternative DNS name matching localhost found.
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:320)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:263)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:258)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:641)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:460)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:360)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1061)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1048)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:995)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.fill(SslConnection.java:642)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:159)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:75)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:151)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:543)
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:398)
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:161)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.security.cert.CertificateException: No subject alternative DNS name matching localhost found.
	at java.base/sun.security.util.HostnameChecker.matchDNS(HostnameChecker.java:207)
	at java.base/sun.security.util.HostnameChecker.match(HostnameChecker.java:98)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:455)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:429)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:283)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:141)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:619)
	... 27 common frames omitted
```

**Comment by joakime** [comment]
In Jetty 9.4.11 the `endpointIdentificationAlgorithm` was unset (aka `null`)
Starting in Jetty 9.4.15.v20190215, the [Endpoint Identification Algorithm was set to `HTTPS`](https://github.com/eclipse/jetty.project/commit/e4d78608a70448fead95bf58057d28590204de47) to bring Jetty's Http Client into line with industry recommendations.

To have similar behavior as 9.4.11 use the following ...

``` java
SslContextFactory.Client sslContextFactory = new SslContextFactory.Client();
sslContextFactory.setEndpointIdentificationAlgorithm(null);

HttpClient client = new HttpClient(sslContextFactory);
```

**Comment by sgrimm** [comment]
Thanks for that workaround! That definitely unblocks me.

This still seems like a bug or at least an unexpected behavior worthy of mention in the documentation, though. SOCKS proxying is effectively a kind of user-space NAT, running at layer 5 rather than layer 3, and shouldn't affect the SSL handshake any more than a layer-3 NAT gateway would.

An administrator shouldn't, I'd think, be able to break (or fix) proxied HTTPS requests just by tweaking the hostname of the proxy.

The commit linked above says the default was changed to prevent problems when `SslContextFactory` was used on a *server*, so maybe the right fix is to only use the new default when `SslContextFactory.Server` is instantiated, and use the old default for `SslContextFactory.Client`.

**Comment by sbordet** [comment]
@sgrimm

there are 2 things happening here: the first is that the SOCKS proxy uses `localhost` as the host name for certificate verification; and the second is that `SslContextFactory.Client` has been strengthened to perform more checks.

Before, the second thing was hiding the first thing.

The reason we changed `SslContextFactory.Client.endpointIdentificationAlgorithm="HTTPS"` by default is to strengthen the security of the client.

Your case is a particular one, where the SOCKS proxy is in between, and it's an orthogonal issue with respect to a more secure `SslContextFactory.Client`.

**Comment by sgrimm** [comment]
Sure, I get that, but my question is more fundamental: Does the current behavior, considering those two things are the case, ever make sense? Will an application developer ever want to connect through a proxy and validate the remote side's SSL certificate using the proxy hostname? Totally understood that this is an interaction between two behaviors, but that doesn't necessarily mean the end result is as intended.

One semi-legitimate use case comes to mind: if a peer has a misconfigured SSL certificate with the wrong hostname, you could use the certificate's hostname for your proxy and fix their misconfiguration. But you could do that by editing your hosts file anyway, so even there, it's not clear to me this behavior is desirable.

I think at the very least, the documentation should mention that SOCKS and SSL don't mix well and describe the workaround.

**Comment by sbordet** [comment]
@sgrimm after the initial bytes to the SOCKS proxy and its reply, the communication is an opaque tunnel - bytes are not interpreted, just forwarded by the proxy to the upstream server and from the server to the client.

The TLS communication is between client and server with the proxy just playing the role of an opaque forwarder, so a client must be able to validate the server certificate as if the proxy was not there - this is to prevent man-in-the-middle attacks.

I'm not sure I understand your semi-legitimate use case. You need a client-side change if your proxy acts as man-in-the-middle: either use a different host name, or have a trust store with a custom CA.

I think it's a bug that we validate against `localhost` rather than the server host name, so I'm investigating what's going on.

**Comment by sgrimm** [comment]
My apologies for being unclear -- I agree this is a bug but I thought maybe your explanation was suggesting that the existing behavior was as intended. Looking forward to this getting fixed; my application uses SOCKS proxies pretty extensively. Thanks for looking into it!

**Comment by sbordet** [comment]
@sgrimm this should be fixed in PR #4394.

Would you be able to test the new code and confirm it fixes your problem?

**Comment by sgrimm** [comment]
Just to confirm: Yes, this fixes the problem. Thanks!
